### ISSUE TITLE
[R-package] 10% CPU when training large dataset

### ISSUE THREAD

**Issue by TpTheGreat** [microsoft/LightGBM#4961]
Hi,
I'm using R.

I'm running on a 72CPU linux machine on AWS.
I am training a 50-100M rows dataset with between 10-120 features.
I have 2 issues:

**1. The CPU of the machine goes up to 10% and not more.**

I went through the documentation and earlier posts with situations like this but unfortunately i couldn't understand much of it.
Could anyone please assist?
I would like the training to go faster and utilize the CPU better.

**the LGBM version installed like this:**
options(repos = c(REPO_NAME = "https://packagemanager.rstudio.com/all/latest"))
install.packages("lightgbm")

**I'm not using num_threads explicitly in the code anywhere.**

**Also not running CV, only training with early stopping on a validation set.**
  

**2. I also predict a large dataset using the same model, predicting a 1K trees model on about 10M rows, takes about 2 minutes.
is there a way to better utilize CPU here?**

Thanks!!

**Comment by jameslamb** [comment]
Thanks for using LightGBM. In the future, please try to follow the issue template when posting here, as that makes it easier for maintainers to understand what's being asked.

Are you able to provide a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) demonstrating this behavior? That would include:

* code you ran (ideally using synthetic or publicly-available data)
* how you're checking CPU utilization (`top`? `htop`?)

I have some theories, but it's hard to say for sure what's happening without more details.

**theory 1: your installation of `{lightgbm}` is not linked against OpenMP**

I ran the following this morning to install `{lightgbm}` the same way you did.

```shell
docker run \
    --rm \
    -v $(pwd):/opt/LightGBM \
    --workdir /opt/LightGBM \
    -it \
    rocker/verse \
    /bin/bash

Rscript -e "install.packages('lightgbm', repos = 'https://packagemanager.rstudio.com/all/latest')"
```

The installed `{lightgbm}` 3.3.1 *from source* (did not match to one of the RStudio Package Manager precompiled binaries). I can tell that OpenMP was linked against LightGBM two ways.

In the installation logs, I see `-fopenmp` passed.

```text
g++ -std=gnu++11 -shared -L/usr/local/lib/R/lib -L/usr/local/lib -o lightgbm.so boosting/boosting.o ...  c_api.o lightgbm_R.o -fopenmp -pthread -L/usr/local/lib/R/lib -lR
```

I can also see it in the list of linked libraries.

```shell
R_HOME=$(Rscript -e "cat(R.home())")
LIGHTGBM_SO=${R_HOME}/site-library/lightgbm/libs/lightgbm.so
ldd ${LIGHTGBM_SO}
```

This line in the output of `ldd` tells me `{lightgbm}` was linked against OpenMP:

```text
libgomp.so.1 => /lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f921e379000)
```

Could you check with these approaches in the environment you're working in? If your find that OpenMP was not linked, then I recommend installing `libomp-dev` and re-installing from source.

**theory 2: environment variable `OMP_NUM_THREADS` is set to a value less than the number of CPUs**

Check the value of environment variable `OMP_NUM_THREADS`. If it is set in your environment, and to a number smaller than the number of CPUs, increase it.

```shell
echo ${OPENMP_NUM_THREADS}
```

See the discussion in #3414 for additional details.

**theory 3: your dataset is simply not large enough, or your model not deep enough, to occupy 100% CPU during training**

This is where it becomes important to see a reproducible example.

**Comment by TpTheGreat** [comment]
Thank you so much @jameslamb !!!
Thanks for a very detailed and thoughtful reply.

- Sorry for not using the template, could you please refer me to where i can find it? (i'm kind of new here)

- I checked and saw that OpenMP is installed correctly + checked that i don't have OMP_NUM_THREADS defined.

- I tried defining the OMP_NUM_THREADS and all of its other similar env variables - but that didn't help.

- After a few more tests, I found that simply adding num_threads to the model hyperparameters, did the trick and now I see more significant CPU usage. still some more things to verify.

- I'll update further if anything new comes up

Thank you again!

**Comment by jameslamb** [comment]
> Sorry for not using the template, could you please refer me to where i can find it? (i'm kind of new here)

I'm referring to all of these sections that were pre-populated when you clicked "new issue" at https://github.com/microsoft/LightGBM/issues.

<img width="884" alt="Screen Shot 2022-02-04 at 9 04 39 PM" src="https://user-images.githubusercontent.com/7608904/152626611-b9df4e15-e0f5-4bd5-b736-424c5855f038.png">

> After a few more tests, I found that simply adding num_threads to the model hyperparameters, did the trick and now I see more significant CPU usage. still some more things to verify.

Oh good! Sorry it took so long for me to respond. Now that's it been a few weeks since your comment...have you verified that adding `num_threads` to the parameters resolves your issue?

There is a proposal in #4972 which, if adopted, would change the R package's behavior to set `num_threads` equal to the number of available CPUs by default.

**Comment by TpTheGreat** [comment]
Thanks @jameslamb !!
